src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,267 @@ PhaseStatus Compiler::optUnrollLoops()
 #pragma warning(pop)
 #endif
 
+//-----------------------------------------------------------------------------
+// optIfConvert
+//
+// Find blocks representing simple if statements represented by conditional jumps
+// over another block. Try to replace the jumps by use of conditional nodes.
+//
+// Arguments:
+//   block -- block that may represent the conditional jump in an if statement.
+//
+// Returns:
+//   true if any IR changes possibly made.
+//
+bool Compiler::optIfConvert(BasicBlock* block)
+{
+#ifndef TARGET_ARM64
+    return false;
+#else
+
+    // Don't optimise the block if it is inside a loop
+    // When inside a loop, branches are quicker than selects.
+    // Detect via the block weight as that will be high when inside a loop.
+    if (block->getBBWeight(this) > BB_UNITY_WEIGHT)
+    {
+        return false;
+    }
+
+    // Does the block end by branching via a JTRUE after a compare?
+    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    {
+        return false;
+    }
+
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* last = block->lastStmt()->GetRootNode();
+    noway_assert(last->gtOper == GT_JTRUE);
+    if (!last->AsOp()->gtOp1->OperIsCmpCompare() || (last->gtFlags & GTF_SIDE_EFFECT) != 0 ||
+        (last->AsOp()->gtOp1->gtFlags & GTF_COLON_COND) != 0)
+    {
+        return false;
+    }
```

Comment:
>GTF_COLON_COND should not be relevant at this point, have you seen issues with it?

I'll try putting it back. Can't quite remember if I was seeing issues with it when the pass was before VN.

>Also, we should be able to get rid of all of these conditions, I see no reason why we cannot create SELECT nodes for these cases. But I am fine with doing that in the future change -- maybe leave a TODO-CQ comment though?

With if conversion you end up always evaluating everything inside the if statement. It seemed important to avoid doing that for things with side conditions.
Eg:
`` if (a==1 && b==foo())``
If a is 1 then you'd avoid executing foo().

(It's possible that GTF_SIDE_EFFECT covers more situations than I was expecting).

Reply:
Yes, I agree for `op1` and `op2` you cannot avoid such checks. But the condition is always evaluated.